immuneML/ml_methods/LSTM.py

```diff
@@ -0,0 +1,300 @@
+
+import datetime
+import random
+import time
+import pyprind
+
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+import keras_tuner as kt
+
+
+from pathlib import Path
+from immuneML.ml_methods.GenerativeModel import GenerativeModel
+from scripts.specification_util import update_docs_per_mapping
+from immuneML.util.PathBuilder import PathBuilder
+
+class LSTM(GenerativeModel):
+
+    def get_classes(self) -> list:
+        pass
+
+    def __init__(self, parameter_grid: dict = None, parameters: dict = None):
+        parameters = parameters if parameters is not None else {}
+        parameter_grid = parameter_grid if parameter_grid is not None else {}
+        super(LSTM, self).__init__(parameter_grid=parameter_grid, parameters=parameters)
+
+        self.checkpoint_dir = ""
+        self.model_params = {}
+        self.char2idx = {}
+        self.idx2char = {}
+        self.max_length = 40
+        self.historydf = None
+        self.generated_sequences = []
+
+    def split_input_target(self, seq):
+        '''
+        split input output, return input-output pairs
+        :param seq:
+        :return:
+        '''
+        input_seq = seq[:-1]
+        target_seq = seq[1:]
+        return input_seq, target_seq
+
+    def _get_ml_model(self, cores_for_training: int = 2, X=None, vocab_size=21, rnn_units=128, embedding_dim=256, batch_size=64):
+
+        #put one hot
+        self.model = tf.keras.Sequential([
+            tf.keras.layers.Embedding(vocab_size,
+                                      embedding_dim,
+                                      batch_input_shape=[batch_size, None]),
+            tf.keras.layers.LSTM(rnn_units,
+                                 return_sequences=True,
+                                 stateful=True,
+                                 recurrent_initializer='glorot_uniform'),
+            tf.keras.layers.Dense(vocab_size)
+        ])
+
+
+    @staticmethod
+    def sample(preds, temperature=1.0):
+        # helper function to sample an index from a probability array
+        preds = np.asarray(preds).astype("float64")
+        preds = np.log(preds) / temperature
+        exp_preds = np.exp(preds)
+        preds = exp_preds / np.sum(exp_preds)
+        probas = np.random.multinomial(1, preds, 1)
+        return np.argmax(probas)
+
+    @staticmethod
+    def loss(labels, logits):
+        '''
+        loss function for the net output
+        :param labels:
+        :param logits:
+        :return:
+        '''
+        loss = tf.keras.losses.sparse_categorical_crossentropy(labels, logits, from_logits=True)
+        return loss
+
+
+    def _fit(self, dataset, cores_for_training: int = 1, result_path: Path = None):
+
+        """
+        Hard values made from previous testing with LSTM. These values may not be suitable for protein sequence
+        generation
+        """
+
+        PathBuilder.build(result_path)
+
+        params = self._parameters
+        alphabet = sorted(set(dataset))
```

Comment:
what does this mean? alphabet information you can get from EnvironmentSettings class depending on the sequence type (amino acid or nucleotide)